File: mip_reference/ruby-sinatra/lib/mip/models/connection.rb
# frozen_string_literal: true

module MIP
  module Models
    # Represents a connection with another MIP node
    class Connection
      STATUSES = %w[PENDING ACTIVE DECLINED REVOKED].freeze

      attr_accessor :status, :daily_rate_limit, :share_my_organization
      attr_reader :mip_identifier, :mip_url, :public_key, :organization_name,
                  :contact_person, :contact_phone, :direction, :created_at,
                  :decline_reason, :revoke_reason

      def initialize(attrs = {})
        @mip_identifier = attrs[:mip_identifier]
        @mip_url = attrs[:mip_url]
        @public_key = attrs[:public_key]
        @organization_name = attrs[:organization_name]
        @contact_person = attrs[:contact_person]
        @contact_phone = attrs[:contact_phone]
        @status = attrs.fetch(:status, 'PENDING')
        @direction = attrs[:direction] # 'inbound' or 'outbound'
        @share_my_organization = attrs.fetch(:share_my_organization, true)
        @daily_rate_limit = attrs.fetch(:daily_rate_limit, 100)
        @created_at = attrs.fetch(:created_at, Time.now.iso8601)
        @decline_reason = attrs[:decline_reason]
        @revoke_reason = attrs[:revoke_reason]
      end

      def active?
        @status == 'ACTIVE'
      end

      def pending?
        @status == 'PENDING'
      end

      def declined?
        @status == 'DECLINED'
      end

      def revoked?
        @status == 'REVOKED'
      end

      def inbound?
        @direction == 'inbound'
      end

      def outbound?
        @direction == 'outbound'
      end

      def approve!(node_profile: nil, daily_rate_limit: 100)
        @status = 'ACTIVE'
        @daily_rate_limit = daily_rate_limit
        update_from_profile(node_profile) if node_profile
      end

      def decline!(reason: nil)
        @status = 'DECLINED'
        @decline_reason = reason
      end

      def revoke!(reason: nil)
        @status = 'REVOKED'
        @revoke_reason = reason
      end

      def restore!
        @status = 'ACTIVE'
        @revoke_reason = nil
      end

      def public_key_fingerprint
        MIP::Crypto.fingerprint(@public_key)
      end

      def to_node_profile
        {
          mip_identifier: @mip_identifier,
          mip_url: @mip_url,
          organization_legal_name: @organization_name,
          contact_person: @contact_person,
          contact_phone: @contact_phone,
          public_key: @public_key,
          share_my_organization: @share_my_organization
        }
      end

      # Create from a connection request payload
      def self.from_request(payload, direction: 'inbound')
        new(
          mip_identifier: payload['mip_identifier'],
          mip_url: payload['mip_url'],
          public_key: payload['public_key'],
          organization_name: payload['organization_legal_name'],
          contact_person: payload['contact_person'],
          contact_phone: payload['contact_phone'],
          share_my_organization: payload.fetch('share_my_organization', true),
          direction: direction,
          status: 'PENDING'
        )
      end

      private

      def update_from_profile(profile)
        @organization_name = profile['organization_legal_name'] if profile['organization_legal_name']
        @contact_person = profile['contact_person'] if profile['contact_person']
        @contact_phone = profile['contact_phone'] if profile['contact_phone']
        @mip_url = profile['mip_url'] if profile['mip_url']
        @public_key = profile['public_key'] if profile['public_key']
      end
    end
  end
end

File: mip_reference/php-symfony/src/Mip/Model/Connection.php
<?php

namespace App\Mip\Model;

use App\Mip\Crypto;

/**
 * Represents a connection with another MIP node
 */
class Connection
{
    public const STATUSES = ['PENDING', 'ACTIVE', 'DECLINED', 'REVOKED'];

    public function __construct(
        public string $mipIdentifier,
        public string $mipUrl,
        public ?string $publicKey,
        public string $organizationName,
        public ?string $contactPerson = null,
        public ?string $contactPhone = null,
        public string $status = 'PENDING',
        public string $direction = 'inbound',
        public bool $shareMyOrganization = true,
        public int $dailyRateLimit = 100,
        public string $createdAt = '',
        public ?string $declineReason = null,
        public ?string $revokeReason = null
    ) {
        if ($this->createdAt === '') {
            $this->createdAt = (new \DateTime())->format(\DateTime::ATOM);
        }
    }

    public function isActive(): bool
    {
        return $this->status === 'ACTIVE';
    }

    public function isPending(): bool
    {
        return $this->status === 'PENDING';
    }

    public function isDeclined(): bool
    {
        return $this->status === 'DECLINED';
    }

    public function isRevoked(): bool
    {
        return $this->status === 'REVOKED';
    }

    public function isInbound(): bool
    {
        return $this->direction === 'inbound';
    }

    public function isOutbound(): bool
    {
        return $this->direction === 'outbound';
    }

    public function approve(?array $nodeProfile = null, int $dailyRateLimit = 100): void
    {
        $this->status = 'ACTIVE';
        $this->dailyRateLimit = $dailyRateLimit;
        if ($nodeProfile !== null) {
            $this->updateFromProfile($nodeProfile);
        }
    }

    public function decline(?string $reason = null): void
    {
        $this->status = 'DECLINED';
        $this->declineReason = $reason;
    }

    public function revoke(?string $reason = null): void
    {
        $this->status = 'REVOKED';
        $this->revokeReason = $reason;
    }

    public function restore(): void
    {
        $this->status = 'ACTIVE';
        $this->revokeReason = null;
    }

    public function getPublicKeyFingerprint(): ?string
    {
        if ($this->publicKey === null) {
            return null;
        }
        return Crypto::fingerprint($this->publicKey);
    }

    public function toNodeProfile(): array
    {
        return [
            'mip_identifier' => $this->mipIdentifier,
            'mip_url' => $this->mipUrl,
            'organization_legal_name' => $this->organizationName,
            'contact_person' => $this->contactPerson,
            'contact_phone' => $this->contactPhone,
            'public_key' => $this->publicKey,
            'share_my_organization' => $this->shareMyOrganization,
        ];
    }

    /**
     * Create from a connection request payload
     */
    public static function fromRequest(array $payload, string $direction = 'inbound'): self
    {
        return new self(
            mipIdentifier: $payload['mip_identifier'],
            mipUrl: $payload['mip_url'],
            publicKey: $payload['public_key'] ?? null,
            organizationName: $payload['organization_legal_name'] ?? 'Unknown',
            contactPerson: $payload['contact_person'] ?? null,
            contactPhone: $payload['contact_phone'] ?? null,
            status: 'PENDING',
            direction: $direction,
            shareMyOrganization: $payload['share_my_organization'] ?? true
        );
    }

    private function updateFromProfile(array $profile): void
    {
        if (!empty($profile['organization_legal_name'])) {
            $this->organizationName = $profile['organization_legal_name'];
        }
        if (!empty($profile['contact_person'])) {
            $this->contactPerson = $profile['contact_person'];
        }
        if (!empty($profile['contact_phone'])) {
            $this->contactPhone = $profile['contact_phone'];
        }
        if (!empty($profile['mip_url'])) {
            $this->mipUrl = $profile['mip_url'];
        }
        if (!empty($profile['public_key'])) {
            $this->publicKey = $profile['public_key'];
        }
    }

    public function toArray(): array
    {
        return [
            'mip_identifier' => $this->mipIdentifier,
            'mip_url' => $this->mipUrl,
            'public_key' => $this->publicKey,
            'organization_name' => $this->organizationName,
            'contact_person' => $this->contactPerson,
            'contact_phone' => $this->contactPhone,
            'status' => $this->status,
            'direction' => $this->direction,
            'share_my_organization' => $this->shareMyOrganization,
            'daily_rate_limit' => $this->dailyRateLimit,
            'created_at' => $this->createdAt,
            'decline_reason' => $this->declineReason,
            'revoke_reason' => $this->revokeReason,
        ];
    }

    public static function fromArray(array $data): self
    {
        return new self(
            mipIdentifier: $data['mip_identifier'],
            mipUrl: $data['mip_url'],
            publicKey: $data['public_key'] ?? null,
            organizationName: $data['organization_name'] ?? 'Unknown',
            contactPerson: $data['contact_person'] ?? null,
            contactPhone: $data['contact_phone'] ?? null,
            status: $data['status'] ?? 'PENDING',
            direction: $data['direction'] ?? 'inbound',
            shareMyOrganization: $data['share_my_organization'] ?? true,
            dailyRateLimit: $data['daily_rate_limit'] ?? 100,
            createdAt: $data['created_at'] ?? '',
            declineReason: $data['decline_reason'] ?? null,
            revokeReason: $data['revoke_reason'] ?? null
        );
    }
}
